src/circulant.jl

```diff
@@ -0,0 +1,122 @@
+export Circulant, CircEig, tocirc
+
+immutable Circulant{T} <: AbstractMatrix{T}
+    c::Vector{T}
+end
+typealias Circ Circulant
+
+function full{T}(C::Circ{T})
+    n = size(C, 1)
+    M = Array(T, n, n) 
+    for i=1:n
+        M[i:n,i] = C.c[1:n-i+1]
+        M[1:i-1,i] = C.c[n-i+2:n]
+    end
+    M
+end
+
+# getindex(C::Circ, i::Int, j::Int) = C.c[mod(i-j,length(C.c))+1]
+isassigned(C::Circ, i::Int, j::Int) = isassigned(C.c,mod(i-j,length(C.c))+1)
+size(C::Circ, r::Int) = (r==1 || r==2) ? length(C.c) :
+    throw(ArgumentError("Invalid dimension $r"))
+size(C::Circ) = size(C,1), size(C,2)
+
+copy(C::Circ) = Circ(C.c)
+conj(C::Circ) = Circ(conj(C.c))
+conj!(C::Circ) = (conj!(C.c); C)
+
+transpose(C::Circ) = Circ(circshift(reverse(C.c), 1))
+ctranspose(C::Circ) = conj!(transpose(C))
+
++(C::Circ, a::Number) = Circ(C.c + a)
++(a::Number, C::Circ) = Circ(a + C.c)
++(C::Circ, D::Circ) = Circ(C.c + D.c)
+-(C::Circ, a::Number) = Circ(C.c - a)
+-(a::Number, C::Circ) = Circ(a - C.c)
+-(C::Circ, D::Circ) = Circ(C.c - D.c)
+-(C::Circ) = Circ(-C.c)
+
+# Return the eigen-factorisation of C. Constituents are efficient to compute
+# and the eigenvectors are represented efficiently using a UnitaryDFT object.
+eig{T<:Real}(C::Circ{T}) = (N = length(C.c); (DFT{Complex{T}}(N) * C.c, UDFT{Complex{T}}(N)))
+eig{T<:Complex}(C::Circ{T}) = (N = length(C.c); (DFT{T}(N) * C.c, UDFT{T}(N)))
+eigfact(C::Circ) = Eigen(eig(C)...)
```

Comment:
This function should convert/promote to the right types such that the answer is a `CircEig`. Currently that is not the case for `Circ{T} where T <: Integer`. That should either be fixed or disallowed.